### PR TITLE
faqs and debugging

### PR DESCRIPTION
**Pull request by melrchen**
based on all customer questions over past 3 months

### CODE REVIEW

File: debugging-errors.mdx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+timeout, so if the query times out before completion, we'd recommend verifying whether the query is still running in the
+console by looking at the query run page and validating that there is a pod in the cluster running the query. Otherwise
+your query may have also completed after the polling timeout, in which case you can also view query status in the
+console.
+
+### Please double check that all features have resolvers and resolvers do not have circular dependencies.
+
+The first thing to validate is that all features have resolvers. If all features have resolvers, then you might see
+this error message if Chalk is unable to construct a dependency graph for resolving the features that you are
+querying. This is usually a result of a join that cannot be processed. Please identify which feature(s) is triggering
```
Comment (AndyMoreland) [inline debugging-errors.mdx:49]: something like -- "we recommend reducing the number of output features in your query until you have a minimal set which still produces this error."

**Review by AndyMoreland — APPROVED** [review]

**Review by michaelwiktorek — APPROVED** [review]

File: debugging-errors.mdx [hunk 2]
```diff
@@ -0,0 +1,120 @@
+---
+title: Debugging
+description: Learn techniques for debugging different elements of Chalk.
+---
+
+We've compiled some of the common errors that people come across, as well as some approaches to debug these errors
+and the most common root causes. If you are still struggling to debug, please reach out in your support channel!
```
Comment (hugokitano) [inline debugging-errors.mdx:7]:
```suggestion
We've compiled some common errors that people come across, as well as some approaches to debug these errors
and their most common root causes. If you are still struggling to debug, please reach out in your support channel!
```

File: debugging-errors.mdx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+If you're receiving either 0 rows when you expect a non-zero query result, or if the feature values that you're seeing
+in your query results do not match what you're expecting, we recommend that you follow the steps in
+[our guide for debugging queries](https://docs.chalk.ai/docs/debugging-queries). This includes some common tips and
+tricks, such as using `store_plan_stages` and stepping through the query plan.
+
+### Unexpected Chalk server error with status code 401
+
+The 401 status code means `unauthenticated`. We'd recommend verifying that the token you're using to run the query is
+correct. To refresh your token, you can also run `chalk login` from the terminal using our CLI and then `chalk config`
```
Comment (hugokitano) [inline debugging-errors.mdx:21]:
```suggestion
correct. To refresh your token, you can run `chalk login` from the terminal using our CLI and then `chalk config`
```

File: debugging-errors.mdx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+to verify that your token has been provisioned for the right environment.
+
+### Unexpected Chalk server error with status code 500
+
+The 500 status code means `internal server error`. This should usually come with another error with more details on
```
Comment (hugokitano) [inline debugging-errors.mdx:26]:
```suggestion
The 500 status code means `internal server error`. This should usually come with another message with more details on
```

File: debugging-errors.mdx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+the root cause of the error. If you are struggling to debug this or find any other clues, please reach out to us!
+
+### Unexpected Chalk server error with status code 502
+
+The 502 status code means `bad gateway`, which usually indicates `timeout` or `not enough servers`. We'd recommend
+looking at CPU utilization on your cluster to determine if you need to modify the cloud resource configurations. You
+can do so in the `Settings > Resources` page in the Chalk console by modifying and applying resource configuration JSON
+changes under `Advanced Resource Configuration`. However, we'd also encourage reaching out to us for discussion on how
+to optimize resource configuration for your situation!
```
Comment (hugokitano) [inline debugging-errors.mdx:35]:
```suggestion
to optimize resource configuration for your environment!
```

File: debugging-errors.mdx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+### Query Timeout / Unclear Failure
+
+If you are running a query from a notebook or somewhere locally, the polling for query status might have a shorter
```
Comment (hugokitano) [inline debugging-errors.mdx:39]:
```suggestion
If you are executing a long-running query from a notebook or somewhere locally, the client's polling for query status might have a shorter
```

File: debugging-errors.mdx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+timeout, so if the query times out before completion, we'd recommend verifying whether the query is still running in the
```
Comment (hugokitano) [inline debugging-errors.mdx:40]: shorter than what?

File: debugging-errors.mdx [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+console by looking at the query run page and validating that there is a pod in the cluster running the query. Otherwise
```
Comment (hugokitano) [inline debugging-errors.mdx]:
```suggestion
console by looking at the query run page and validating that there is a pod in the cluster running the query. It's possible that
```

File: debugging-errors.mdx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+your query may have also completed after the polling timeout, in which case you can also view query status in the
+console.
```
Comment (hugokitano) [inline debugging-errors.mdx]:
```suggestion
dashboard.
```

File: debugging-errors.mdx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+can do so in the `Settings > Resources` page in the Chalk console by modifying and applying resource configuration JSON
+changes under `Advanced Resource Configuration`. However, we'd also encourage reaching out to us for discussion on how
+to optimize resource configuration for your situation!
+
+### Query Timeout / Unclear Failure
+
+If you are running a query from a notebook or somewhere locally, the polling for query status might have a shorter
+timeout, so if the query times out before completion, we'd recommend verifying whether the query is still running in the
+console by looking at the query run page and validating that there is a pod in the cluster running the query. Otherwise
+your query may have also completed after the polling timeout, in which case you can also view query status in the
```
Comment (hugokitano) [inline debugging-errors.mdx:42]:
```suggestion
your query may also complete after the polling timeout, in which case you can view the query status in the
```

File: debugging-errors.mdx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+console.
+
+### Please double check that all features have resolvers and resolvers do not have circular dependencies.
+
+The first thing to validate is that all features have resolvers. If all features have resolvers, then you might see
```
Comment (hugokitano) [inline debugging-errors.mdx:47]: specifically, all features need a resolver whose output includes them. idk how to say in a non wordy way

File: debugging-errors.mdx [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+this error message if Chalk is unable to construct a dependency graph for resolving the features that you are
+querying. This is usually a result of a join that cannot be processed. Please identify which feature(s) is triggering
+the circular dependency error and reach out to us with this context!
+
+### Duplicate Resolver Shortname
+
+The first thing to check is whether there actually are duplicate resolvers by the same name. If there are not, this
+error message can also sometimes be an indicator of a missing import.
```
Comment (hugokitano) [inline debugging-errors.mdx:56]: how would they search for how to solve this?

File: debugging-errors.mdx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+from chalk import ChalkClient
+
+if __name__=='__main__':
+    ChalkClient().offline_query(
+        input={User.id: [1]},
+        output=[User.sum_transactions],
+        store_plan_stages=True,
+        recompute_features=True
+    ).resolver_replay(sum_transaction_amount)
+```
```
Comment (hugokitano) [inline debugging-errors.mdx:99]: I have PR out to make this impossible btw, so maybe hold off on this particular reason

File: debugging-errors.mdx [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+### Could not connect to server when running `chalk apply`
+
+If you've verified that you have a token and are properly authenticated, then you should also verify what role you
+have within the Chalk environment. The `developer` role can only run `chalk apply --branch` rather than `chalk apply`.
```
Comment (hugokitano) [inline debugging-errors.mdx:104]:
```suggestion
have within the Chalk environment. The `developer` role can run `chalk apply --branch` but not `chalk apply`.
```

File: debugging-errors.mdx [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+## Deployed Errors
+
+These are some errors that you might see in features/resolvers that you've deployed that are fully orchestrated,
+as opposed to features/resolvers that you might be actively iterating on.
```
Comment (hugokitano) [inline debugging-errors.mdx:109]: i think this is pretty vague as to what is actually happening and how to debug

File: debugging-errors.mdx (hunk 13 above)
Comment (melrchen) [reply debugging-errors.mdx:99]: omg ok holding

**Pull request merged**